Fixed footprints of several RA and TD base structures

reaperrr opened the pull request:
There were several cases of base buildings in RA and TD with the top row of the footprint only consisting of `_` cells, namely

RA:
- Advanced Power Plant
- Fake Advanced Power Plant
- Tesla Coil
- Gap Generator
- Flak Cannon

TD:
- Weapons Factory
- Temple of Nod
- Hand of Nod
- Obelisk
- Advanced Guard Tower

In the case of defenses, this meant that the `CenterPosition` was located between the empty cell and the occupied cell, rather than centered on the occupied cell, which - while consistent with TD - was not consistent with original RA and made placing, attacking or C4'ing these structures look a bit dodgy.

To make placement, attacking, C4'ing and later on HitShape setup easier and more consistent, I've removed all those empty top rows and adjusted the art offsets instead.

_Note_: Due to the sheer amount of actors that would be affected, I've refrained from doing the same for trees and rocks (fortunately civilian structures are not affected), since they are indestructible in RA/TD and therefore wouldn't really benefit from similar fixes. And, of course, the diff and risk of regressions would go up considerably.

_Note2_: The upgrade rule did work on two maps, but then I ran into a bug in our map upgrading so I had to use a slightly modified version of my previous MapFormat upgrade rule to apply it to all maps.
Additionally, I ran into a zlib error when trying to upgrade the `.oramap` packages, so I could only fix the unzipped maps (though I'm not sure we ship any packaged maps which have these specific actors pre-placed.)

reaperrr committed:
Tweaked RA airfield + soviet barracks offsets

Barracks art is now more centered and aligned with allied tent. Airfield is now more centered, which also makes the custom spawn position obsolete and fixes the misalignment between spawn and landing position.

This is also done to reduce misalignment between art and footprint-sized hit shapes that will follow later.

reaperrr committed:
Removed empty top rows from RA base structure footprints

Removed top row of buildings where it only consisted of _ cells and adjusted the art offset instead. This fixes the issue that CenterPosition was located in the middle between the occupied and passable cell for defenses like tesla coils, makes their placement a bit more intuitive and makes setting up proper HitShapes easier.

reaperrr committed:
Removed empty top rows from TD base structure footprints

Removed top row of buildings where it only consisted of _ cells and adjusted the art offset instead. This fixes the issue that CenterPosition was located in the middle between the occupied and passable cell for defenses like Obelisks, makes their placement a bit more intuitive and makes setting up proper HitShapes easier.

reaperrr committed:
Map actor Location upgrade rules

To move all buildings with changed footprint down by 1 cell.

reaperrr committed:
Fixed map placement of RA buildings

...with changed footprint.

reaperrr committed:
Fixed map placement of TD buildings

...with changed footprint.

pchote commented:
This will cause issues for the map importer, right? (IIRC we already have that problem for tib trees and maybe other things).

Can you add something similar to the upgrade rules to those too?

Biofreak1987 commented:
A few lua files need to be updated as well as they include coordinates for rebuilding lost structures... though I might do that with a different pr as I've planned to rework some of the AI code.

reaperrr committed:
Add support for custom locations to legacy map importer

For example, to place actors with changed footprint correctly.
Use it to ensure correct positioning of several RA and TD structures as well as TD tiberium trees.

reaperrr committed:
Fix TD mission base rebuild locations

To adapt to changed footprint.

reaperrr commented:
Rebased, added "upgrade rule" to importer (for TD tib trees as well), and updated AI build positions for TD weapons factory (no other cases in TD or RA, as far as I could tell).

obrakmann commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -934,7 +934,144 @@ internal static void UpgradeActors(int engineVersion, ref List<MiniYamlNode> nod
+				if (engineVersion < 20160530 && Game.ModData.Manifest.Mod.Id == "ra")
+				{
+					// Tesla Coil
+					if (node.Value.Value == "tsla")
```
I don't quite see the point of these upgrade rules. With hard-coding things like this, it'll only work on the official ra mod, which - when you have the utility that contains these rules - already has been adapted. I think you can safely remove these, unless I'm missing something not obvious.

reaperrr replied:
This is for upgrading maps from the resource center that don't ship with OpenRA, mostly.
@pchote requested them (iirc).

pchote replied:
This is indeed for the resource center and other third party maps.  It sucks to have these checks hardcoded in common code, but we can't fix that until we redo the upgrade rules plumbing.

pchote commented:
> Note: Due to the sheer amount of actors that would be affected, I've refrained from doing the same for trees and rocks (fortunately civilian structures are not affected), since they are indestructible in RA/TD and therefore wouldn't really benefit from similar fixes. And, of course, the diff and risk of regressions would go up considerably.

Do #11180 and #11377 change this?

pchote commented on OpenRA.Mods.Common/UtilityCommands/ImportLegacyMapCommand.cs:
```diff
@@ -415,6 +417,10 @@ public static void LoadActors(IniFile file, string section, List<string> players
 
 		public abstract string ParseTreeActor(string input);
 
+		public virtual int ModifyLocationX(string input) { return 0; }
```
Can we please change this to a `public virtual CPos ParseActorLocation(int loc)` and include the `mapSize` calculation here?

pchote replied:
Alternatively you could pull out a `ParseActor` (like `ParseTreeActor` below) and then move the `LocationInit` stuff into both of these.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -934,7 +934,144 @@ internal static void UpgradeActors(int engineVersion, ref List<MiniYamlNode> nod
 			{
-				// Add rules here
+				// Fix RA building footprints to not use _ when it's not necessary
+				if (engineVersion < 20160530 && Game.ModData.Manifest.Mod.Id == "ra")
```
Can you please pass `ModData modData` in as the first argument to these methods?  Referencing `Game.ModData` causes a bunch of issues, and i'm slowly trying to eliminate all these references.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -934,7 +934,144 @@ internal static void UpgradeActors(int engineVersion, ref List<MiniYamlNode> nod
+					// Tesla Coil
+					if (node.Value.Value == "tsla")
+					{
+						var locationNode = node.Value.Nodes.First(n => n.Key == "Location");
```
Lots of code duplication here.  Can you pull out a common method to do this?

pchote replied:
I think the cleanest solution would be to define a `Dictionary<string, CVec> locationOffsets` for each mod and then do the manipulation if it `ContainsKey(node.Value.Value)`.

reaperrr commented:
> Do #11180 and #11377 change this?

Not really. Burning trees are just a cosmetical feature, and all flame weapons have enough spread that we don't need footprint-shaped HitShapes for trees to make sure they get hit by them.

reaperrr committed:
Pass ModData to upgrade methods

reaperrr commented:
Updated.

I didn't entirely do as requested for the upgrade rule, but mostly because I wasn't sure how exactly to implement the suggestions so I opted for a compromise that is hopefully sufficient.

pchote commented:
50fbe66f05ef3fd3a03cf9e82acf9c537c37fae3 looks good to me.

Mailaender commented:
This needs to take https://github.com/OpenRA/OpenRA/pull/11368 into account as well now.

Biofreak1987 commented:
gdi07 is missing as well

reaperrr commented:
Rebased and adapted Nod7c + GDI7.

Mailaender commented:
Can't spot regressions. :white_check_mark:

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -182,61 +182,94 @@ internal static void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode
+					if (buildings.Contains(node.Value.Value))
+					{
+						var locationNode = node.Value.Nodes.First(n => n.Key == "Location");
+						var parts = locationNode.Value.Value.Split(',');
```
You should be able to use `FieldLoader.GetValue<CPos>("Location", locatioinNode.Value.Value)` instead of rolling your own parser.

pchote replied:
It might be a good idea to extract out a method that takes a ref string and a CVec offset and does the manipulation in place.  You can then use it here and below.

pchote replied:
See https://github.com/OpenRA/OpenRA/blob/0ac7b73d37edd18d13efb9e8a9d5ecc9be7e711f/OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs#L69

reaperrr commented:
Rebased and updated.

I can't test the refactored upgrade rule, so review the code carefully. I'll squash the two commits together if the last commit is considered fine.

For the record, the Travis and AppVeyor failures are both not the the fault of this PR.

reaperrr commented:
GH messed up the commit order again. https://github.com/OpenRA/OpenRA/pull/11373/commits/31e008e4ec540224b852bdf49e98b70469994c80 is actually the 2nd-to-last commit, right before the last one.

reaperrr commented:
Updated.

pchote commented on mods/ra/rules/structures.yaml:
```diff
@@ -408,13 +408,13 @@ TSLA:
 	Building:
-		Footprint: _ x
-		Dimensions: 1,2
+		Footprint: x
```
This is the default footprint, so you can remove it completely (here and the other defenses).

pchote commented on mods/ra/sequences/structures.yaml:
```diff
@@ -131,12 +131,16 @@ powr:
 
 apwr:
 	idle:
+		Offset: 0,-10
```
Shouldn't these be -12 (half a cell)?

pchote replied:
You mention in one of the other commits that you are also adjusting for future footprint positions, so I guess this may be intentional.

reaperrr replied:
Yeah, this change is basically similar to the airfield and barracks offset changes, it's supposed to make the visuals match the occupied cells a little better.

pchote commented on mods/cnc/maps/nod07a/nod07a-AI.lua:
```diff
@@ -16,7 +16,7 @@ BaseNuke1 = { type = "nuke", pos = CPos.New(16, 56), cost = 500, exists = true }
 BaseNuke3 = { type = "nuke", pos = CPos.New(27, 51), cost = 500, exists = true }
 InfantryProduction = { type = "pyle", pos = CPos.New(18, 54), cost = 500, exists = true }
-VehicleProduction = { type = "weap", pos = CPos.New(27, 54), cost = 2000, exists = true }
+VehicleProduction = { type = "weap", pos = CPos.New(27, 55), cost = 2000, exists = true }
```
Not for this PR, but it would be nice if we could place these in the map editor and query this info at setup time.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -198,61 +198,83 @@ internal static void UpgradeWeaponRules(int engineVersion, ref List<MiniYamlNode
-				// Add rules here
-				UpgradeActors(engineVersion, ref node.Value.Nodes, node, depth + 1);
+				// Fix RA building footprints to not use _ when it's not necessary
+				if (engineVersion < 20160530 && modData.Manifest.Mod.Id == "ra" && depth == 2)
```
Can you please bump this date up to current?

pchote replied:
These need to be `depth == 1` if you want the rule to run :smile:

reaperrr replied:
> These need to be depth == 1 if you want the rule to run

Hm... are you sure? I thought using `parent` for the actor node (which should be depth 1) and `node` for the `Location` (should be depth 2) would be correct.

pchote replied:
When this runs the actor node is depth 0 and `Location` etc are depth 1.

reaperrr replied:
Alright.

pchote commented on mods/cnc/rules/structures.yaml:
```diff
@@ -821,12 +822,12 @@ ATWR:
-		Footprint: _ x
-		Dimensions: 1,2
+		Footprint: x
+		Dimensions: 1,1
```
Don't forget about the firing offsets!

reaperrr replied:
In case of the ATWR, I changed the 'turret' offset instead.

pchote replied:
When I tested this the missiles came from tabove the roof instead of from the windows or the launchers below the windows.

pchote commented:
You've missed RA's evacuation map (merged after you started on this?).

Aside from that and the bugs above, the fixed upgrade rules match the changes you've done here.

reaperrr commented:
Updated.

- removed redundant default footprints from yaml
- removed redundant default spawn offset from RA airfield
- lowered TD `ATWR` firing offset
- fixed date and depth of upgrade rule
- adapted RA Evacuation

pchote commented:
Unfortunately this has regressed the selection rectangles on at least the weapons factory, temple, and hand.  I haven't checked the other structures:

![screen shot 2016-06-19 at 18 20 33](https://cloud.githubusercontent.com/assets/167819/16178798/c06a5e54-364a-11e6-9baa-29d76c0a65e8.png)

reaperrr committed:
Fix TD selection and section deco offsets

...of Temple of Nod, Weapons Factory and Hand of Nod, to better match the new footprints.

reaperrr committed:
Fix RA selection box offsets

Gap Gen, Adv. Powerplant, fake APP, Tesla Coil.

reaperrr commented:
Updated, fixed selection boxes of those 3. I did take a look at the other structures as well, but the defenses look fine to me and the selection box of the apwr/fake apwr looked just slightly too small, which I adressed just in case.

reaperrr commented:
![td-select](https://cloud.githubusercontent.com/assets/2857877/16179139/a084da8e-365d-11e6-8c6c-b93ed4904eee.png)
![ra-select](https://cloud.githubusercontent.com/assets/2857877/16179143/a403e830-365d-11e6-8a9f-c8e72751f4cd.png)

reaperrr commented:
As requested on IRC, tweaked Gap Gen and Tesla selection boxes:
![gap-tsla--fixed](https://cloud.githubusercontent.com/assets/2857877/16179222/897b7b48-365f-11e6-8c35-c29272b7dfa1.png)

reaperrr commented:
Updated. I hope this is it...

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

Not merging, as I'm not sure if @Mailaender's ✅ is still valid.

pchote commented:
Needs a rebase now.

reaperrr commented:
Rebased. Travis failed at something unrelated.

Mailaender commented:
I just checked infantry entering buildings as I thought this was the point of the pull request. It looks like @pchote checked this more thoroughly and we should wait for this final. :+1:

obrakmann commented:
hm, you didn't touch the Iron Curtain, did you? It gets imported with a wrong offset as well:
![screenshot from 2016-06-21 21 58 35](https://cloud.githubusercontent.com/assets/4331210/16244354/d41e7ab8-37fb-11e6-80ce-ac91e3fad238.png). It should look like this: 
![allies14](https://cloud.githubusercontent.com/assets/4331210/16244413/25405722-37fc-11e6-882e-c1384daee78c.png)

obrakmann commented on OpenRA.Mods.RA/ImportRedAlertLegacyMapCommand.cs:
```diff
@@ -126,6 +126,17 @@ public override string ParseTreeActor(string input)
+			var newLoc = new CPos(loc % MapSize, loc / MapSize);
+			var vectorDown = new CVec(0, 1);
+
+			if (input == "tsla" || input == "agun" || input == "gap" || input == "apwr")
```
OK, so it looks like this should just contain `iron` as well. No need to add it to the upgrade rules since its rules haven't changed.

obrakmann commented:
:+1: after that has been addressed. Haven't checked all the maps, but those that I took at look at were fine.

reaperrr commented:
Updated. Added Iron Curtain to RA map importer location modifier.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a07d045378564146371aea5252c59783e2fb21a6)
